x/wasm/dynamic_link_test.go

```diff
@@ -0,0 +1,334 @@
+package wasm
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/line/lbm-sdk/x/wasm/keeper"
+	abci "github.com/line/ostracon/abci/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	calleeContract     = mustLoad("./testdata/dynamic_callee_contract.wasm")
+	callerContract     = mustLoad("./testdata/dynamic_caller_contract.wasm")
+	numberContract     = mustLoad("./testdata/number.wasm")
+	callNumberContract = mustLoad("./testdata/call_number.wasm")
+)
+
+// This tests dynamic calls using callee_contract's pong
+func TestDynamicPingPongWorks(t *testing.T) {
+	// setup
+	data := setupTest(t)
+
+	h := data.module.Route().Handler()
+
+	// store dynamic callee code
+	storeCalleeMsg := &MsgStoreCode{
+		Sender:       addr1,
+		WASMByteCode: calleeContract,
+	}
+	res, err := h(data.ctx, storeCalleeMsg)
+	require.NoError(t, err)
+
+	calleeCodeId := uint64(1)
+	assertStoreCodeResponse(t, res.Data, calleeCodeId)
+
+	// store dynamic caller code
+	storeCallerMsg := &MsgStoreCode{
+		Sender:       addr1,
+		WASMByteCode: callerContract,
+	}
+	res, err = h(data.ctx, storeCallerMsg)
+	require.NoError(t, err)
+
+	callerCodeId := uint64(2)
+	assertStoreCodeResponse(t, res.Data, callerCodeId)
+
+	// instantiate callee contract
+	instantiateCalleeMsg := &MsgInstantiateContract{
+		Sender:  addr1,
+		CodeID:  calleeCodeId,
+		Label:   "callee",
+		InitMsg: []byte(`{}`),
+		Funds:   nil,
+	}
+	res, err = h(data.ctx, instantiateCalleeMsg)
+	require.NoError(t, err)
+
+	calleeContractAddress := parseInitResponse(t, res.Data)
+
+	// instantiate caller contract
+	cosmwasmInstantiateCallerMsg := fmt.Sprintf(`{"callee_addr":"%s"}`, calleeContractAddress)
+	instantiateCallerMsg := &MsgInstantiateContract{
+		Sender:  addr1,
+		CodeID:  callerCodeId,
+		Label:   "caller",
+		InitMsg: []byte(cosmwasmInstantiateCallerMsg),
+		Funds:   nil,
+	}
+	res, err = h(data.ctx, instantiateCallerMsg)
+	require.NoError(t, err)
+
+	callerContractAddress := parseInitResponse(t, res.Data)
+
+	// execute ping
+	cosmwasmExecuteMsg := `{"ping":{"ping_num":"100"}}`
+	executeMsg := MsgExecuteContract{
+		Sender:   addr1,
+		Contract: callerContractAddress,
+		Msg:      []byte(cosmwasmExecuteMsg),
+		Funds:    nil,
+	}
+	res, err = h(data.ctx, &executeMsg)
+	require.NoError(t, err)
+
+	assert.Equal(t, len(res.Events), 3)
+	assert.Equal(t, "wasm", res.Events[0].Type)
+	assert.Equal(t, len(res.Events[0].Attributes), 6)
+	assertAttribute(t, "contract_address", "link10pyejy66429refv3g35g2t7am0was7yaducgya", res.Events[0].Attributes[0])
```

Comment:
And this address seems to be repeating. If so, how about changing to a const property?

Reply:
The contract address is always the same because the code id and instance id are the same. (https://github.com/line/lbm-sdk/blob/a18b7c8c01a77e5a1290a0a4600feda80ad3abce/x/wasm/keeper/keeper.go#L878-L882) 

But, it is out of this test, so I will delete this asserting.